Add Kafka::Librd::producev with header support

allsopp opened the pull request.

Commit:
Bump Alien::Librdkafka

Commit:
Add Kafka::Librd::producev with header support

This new method wraps the rd_kafka_producev function, introduced in
librdkafka v0.11.4[1], which supports adding Kafka record headers[2].

[1] https://github.com/confluentinc/librdkafka/releases/tag/v0.11.4
[2] https://cwiki.apache.org/confluence/display/KAFKA/KIP-82+-+Add+Record+Headers

Commit:
Fix bump to use correct version format

edmundadjei commented on lib/Kafka/Librd.pm:
```diff
@@ -173,6 +173,45 @@ If an error occurs during creation of the topic, C<undef> is returned. In such
 
+=head2 producev
+
+    $err = $topic->producev(\%params)
```
$kafka->producev ?

allsopp replied:
Good spot! Fixed.

edmundadjei commented on t/producev.t:
```diff
@@ -0,0 +1,81 @@
+
+BEGIN { use_ok "Kafka::Librd" }
+my $krd = Kafka::Librd->new(
+    Kafka::Librd::RD_KAFKA_CONSUMER,
```
RD_KAFKA_CONSUMER? do we have RD_KAFKA_PRODUCER we can use?

allsopp replied:
Yeah, we have that. From `rdkafka.h`:

```
typedef enum rd_kafka_type_t {
	RD_KAFKA_PRODUCER, /**< Producer client */
	RD_KAFKA_CONSUMER  /**< Consumer client */
} rd_kafka_type_t;
```

So it would be correct to use `RD_KAFKA_PRODUCER` here. Updated! :smile:

edmundadjei commented on t/producev.t:
```diff
@@ -0,0 +1,81 @@
+        msgflags  => 0,
+        key       => "key",
+        value     => "value",
+        headers   => { foo => 1, },
```
are values in headers restricted to scalars?  will { foo => { bar => baz, } },  work for instance?

allsopp replied:
There is no restriction but Perl will coerce a reference to a string. So, in your example the headers would be e.g. 

```
    headers     {
        foo => "HASH(0x5ea8d52a4600)"
    },
```

edmundadjei replied:
maybe we should add a note to that effect

allsopp replied:
Added

edmundadjei commented on Rdkafka.xs:
```diff
@@ -288,6 +288,74 @@ krd_flush(rdk, timeout_ms)
+int
+krd_producev(rdk, params)
+        rdkafka_t* rdk
+        HV* params
```
can we add timestamp and default it to one of the macros? RD_KAFKA_TIMESTAMP_CREATE_TIME maybe?

allsopp replied:
Something, either `librdkafka` or the broker, is already adding the current time as the `timestamp` with the type `RD_KAFKA_TIMESTAMP_CREATE_TIME` which corresponds to "Message creation time". I confirmed this by consuming a message produced with `librdkafka`. I'm not sure whether it's the library or the broker adding this, although I would presume `RD_KAFKA_TIMESTAMP_CREATE_TIME` means it's the library. I will investigate.

allsopp replied:
It's the library in `rdkafka_msg.c`:

```
        if (timestamp)
                rkm->rkm_timestamp = timestamp;
        else
                rkm->rkm_timestamp = rd_uclock() / 1000;
        rkm->rkm_tstype = RD_KAFKA_TIMESTAMP_CREATE_TIME;
```

edmundadjei replied:
I believe the broker adds it, as $msg->timestamp always has a valuue. This is only if we want to send a custom value when we produce messages. RD_KAFKA_TIMESTAMP_CREATE_TIME is probably the most reasonable default

edmundadjei replied:
In any case this request is not a blocker

edmundadjei commented on Rdkafka.xs:
```diff
@@ -288,6 +288,74 @@ krd_flush(rdk, timeout_ms)
+            value = NULL; vlen = 0;
+        }
+
+        hdrs = rd_kafka_headers_new(0);
```
how do we handle malloc failures here?

allsopp replied:
As far as I can tell, the library has its own [`malloc`](https://github.com/confluentinc/librdkafka/blob/master/src/rd.h#L139) that calls `assert` from the C library on failure.

edmundadjei replied:
yeah, just noticed

edmundadjei commented on Rdkafka.xs:
```diff
@@ -288,6 +288,74 @@ krd_flush(rdk, timeout_ms)
+        hsv = hv_fetchs(params, "headers", 0);
+        if (hsv != NULL && SvOK(*hsv)) {
+            if (!SvROK(*hsv) || SvTYPE(SvRV(*hsv)) != SVt_PVHV)
+                croak("headers must be a hash reference");
```
do we need to free memory before we croak? I noticed "rd_kafka_headers_destroy" in the C lib

allsopp replied:
The docs say:

> The message object will assume ownership of the headers (unless producev() * fails).

Although it's not clear what fails means in this context. I'll investigate.

edmundadjei replied:
the monolith daemon keeps crashing, and I'm wondering if we have memory leaks

allsopp replied:
I've updated the code to destroy the header object when `producev` fails.

Commit:
Destroy headers on rd_kafka_producev failure

According to the documentation, the message object takes ownership of
the headers unless rd_kafka_producev fails [1], so explicitly destroy
the headers in the case of an error.

Also pass the number of headers to rd_kafka_headers_new up front, so the
library doesn't need to reallocate later.

[1] https://github.com/confluentinc/librdkafka/blob/v1.3.0/src/rdkafka.h#L1029

allsopp commented:
Internally approved. Will raise a PR upstream and sync back to this fork if and when accepted.
